enhancements/ingress/set-replace-append-delete-deny-headers.md

```diff
@@ -0,0 +1,491 @@
+    http-response del-header X-XSS-Protection: 1;mode=block    
+```
+Use of a single environment variable `ROUTER_HTTP_HEADERS` for both action `Set` and `Delete` will help preserve ordering.
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting or Deleting HTTP headers via Route's Spec by Users.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+* openshift/openshift-apiserver
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `SetHTTPResponseHeaders` or `DeleteHTTPResponseHeaders` based on the action `Set` or `Delete` respectively, which specify the response headers to be set or deleted in the
+the haproxy.config sections corresponding to the following backends: plain http backend, backend with TLS terminated at the edge, secure backend with re-encryption ( [code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L609)).
+The router sets this variable based on the Route's `spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+```go
+// ServiceAliasConfig is a route for a service.  Uniquely identified by host + path.
+type ServiceAliasConfig struct {
+    ...
+	// For route based setting of headers
+	SetHTTPResponseHeaders              []SetHTTPHeader
+	DeleteHTTPResponseHeaders           []DeleteHTTPHeader
+}
+```
+```go
+// SetHTTPHeader specifies an HTTP header that should be set.
+type SetHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+
+	// Value specifies the header value.
+	Value string
+
+	// Action specifies the action to be performed.
+	Action string
+}
+
+type DeleteHTTPHeader struct {
+	// Name specifies an HTTP header name.
+	Name string
+}
+```
+
+In this case the ordering of actions will be maintained as we will be having access to the RouteSpec inside router. So, we will run a loop
+on either `spec.httpHeaders.headersManipulation.response[*].set` or `spec.httpHeaders.headersManipulation.response[*].delete` which will have the desired
+sequence.
+
+C) We will be creating an immutable/harcoded deny list of HTTP Response headers in the [library-go](https://github.com/openshift/library-go/blob/2a52b7ce52f99e2ed185718380b43734877f2439/pkg/route/validation/validation.go#L48-L141) 
```

Comment:
If you do this, you will need to make sure that is documented in the godoc for this new API, I'm not sure I agree with the approach though, how many different headers are going to be blocked?

Reply:
For now we are blocking the HSTS header only as we already have a way to set this via route annotation.